Cycle detected on node type editor.collection/CollectionNode

Exception:

```
{:cause
 "Assert failed: Cycle detected on node type editor.collection/CollectionNode and output :build-targets\n(not (contains? (:in-production ctx) [node-id label]))",
 :via
 [{:type java.util.concurrent.ExecutionException,
   :message
   "java.lang.AssertionError: Assert failed: Cycle detected on node type editor.collection/CollectionNode and output :build-targets\n(not (contains? (:in-production ctx) [node-id label]))",
   :at [java.util.concurrent.FutureTask report "FutureTask.java" 122]}
  {:type java.lang.AssertionError,
   :message
   "Assert failed: Cycle detected on node type editor.collection/CollectionNode and output :build-targets\n(not (contains? (:in-production ctx) [node-id label]))",
   :at [internal.node$in_production invokeStatic "node.clj" 1280]}],
 :trace
 [[internal.node$in_production invokeStatic "node.clj" 1280]
  [internal.node$in_production invoke "node.clj" 1279]
  [editor.collection$CollectionNode$behavior$build_targets invokeStatic "collection.clj" 419]
  [editor.collection$CollectionNode$behavior$build_targets invoke "collection.clj" 419]
  [clojure.lang.Var invoke "Var.java" 383]
  [internal.node.NodeImpl produce_value "node.clj" 268]
  [internal.node$pull_first_input_value invokeStatic "node.clj" 1093]
  [internal.node$pull_first_input_value invoke "node.clj" 1088]
  [editor.factory$FactoryNode$behavior$build_targets invokeStatic "factory.clj" 86]
  [editor.factory$FactoryNode$behavior$build_targets invoke "factory.clj" 86]
  [clojure.lang.Var invoke "Var.java" 383]
  [internal.node.NodeImpl produce_value "node.clj" 268]
  [internal.node$pull_first_input_value invokeStatic "node.clj" 1093]
  [internal.node$pull_first_input_value invoke "node.clj" 1088]
  [editor.game_object$ReferencedComponent$behavior$build_targets invokeStatic "game_object.clj" 143]
  [editor.game_object$ReferencedComponent$behavior$build_targets invoke "game_object.clj" 143]
  [clojure.lang.Var invoke "Var.java" 383]
  [internal.node.NodeImpl produce_value "node.clj" 268]
  [internal.node$pull_input_values$fn__21050 invoke "node.clj" 1109]
  [clojure.core$mapv$fn__6953 invoke "core.clj" 6627]
  [clojure.lang.PersistentVector reduce "PersistentVector.java" 341]
  [clojure.core$reduce invokeStatic "core.clj" 6544]
  [clojure.core$mapv invokeStatic "core.clj" 6618]
  [clojure.core$mapv invoke "core.clj" 6618]
  [internal.node$pull_input_values invokeStatic "node.clj" 1107]
  [internal.node$pull_input_values invoke "node.clj" 1104]
  [editor.game_object$GameObjectNode$behavior$build_targets invokeStatic "game_object.clj" 278]
  [editor.game_object$GameObjectNode$behavior$build_targets invoke "game_object.clj" 278]
  [clojure.lang.Var invoke "Var.java" 383]
  [internal.node.NodeImpl produce_value "node.clj" 268]
  [internal.node$pull_first_input_value invokeStatic "node.clj" 1093]
  [internal.node$pull_first_input_value invoke "node.clj" 1088]
  [editor.collection$ReferencedGOInstanceNode$behavior$build_targets invokeStatic "collection.clj" 262]
  [editor.collection$ReferencedGOInstanceNode$behavior$build_targets invoke "collection.clj" 262]
  [clojure.lang.Var invoke "Var.java" 383]
  [internal.node.NodeImpl produce_value "node.clj" 268]
  [internal.node$pull_input_values$fn__21050 invoke "node.clj" 1109]
  [clojure.core$mapv$fn__6953 invoke "core.clj" 6627]
  [clojure.lang.PersistentVector reduce "PersistentVector.java" 341]
  [clojure.core$reduce invokeStatic "core.clj" 6544]
  [clojure.core$mapv invokeStatic "core.clj" 6618]
  [clojure.core$mapv invoke "core.clj" 6618]
  [internal.node$pull_input_values invokeStatic "node.clj" 1107]
  [internal.node$pull_input_values invoke "node.clj" 1104]
  [editor.collection$CollectionNode$behavior$build_targets invokeStatic "collection.clj" 419]
  [editor.collection$CollectionNode$behavior$build_targets invoke "collection.clj" 419]
  [clojure.lang.Var invoke "Var.java" 383]
  [internal.node.NodeImpl produce_value "node.clj" 268]
  [internal.node$pull_first_input_value invokeStatic "node.clj" 1093]
  [internal.node$pull_first_input_value invoke "node.clj" 1088]
  [editor.collection$CollectionInstanceNode$behavior$build_targets invokeStatic "collection.clj" 492]
  [editor.collection$CollectionInstanceNode$behavior$build_targets invoke "collection.clj" 492]
  [clojure.lang.Var invoke "Var.java" 383]
  [internal.node.NodeImpl produce_value "node.clj" 268]
  [internal.node$pull_input_values$fn__21050 invoke "node.clj" 1109]
  [clojure.core$mapv$fn__6953 invoke "core.clj" 6627]
  [clojure.lang.PersistentVector reduce "PersistentVector.java" 341]
  [clojure.core$reduce invokeStatic "core.clj" 6544]
  [clojure.core$mapv invokeStatic "core.clj" 6618]
  [clojure.core$mapv invoke "core.clj" 6618]
  [internal.node$pull_input_values invokeStatic "node.clj" 1107]
  [internal.node$pull_input_values invoke "node.clj" 1104]
  [editor.collection$CollectionNode$behavior$build_targets invokeStatic "collection.clj" 419]
  [editor.collection$CollectionNode$behavior$build_targets invoke "collection.clj" 419]
  [clojure.lang.Var invoke "Var.java" 383]
  [internal.node.NodeImpl produce_value "node.clj" 268]
  [internal.node$pull_input_values$fn__21050 invoke "node.clj" 1109]
  [clojure.core$mapv$fn__6953 invoke "core.clj" 6627]
  [clojure.lang.PersistentVector reduce "PersistentVector.java" 341]
  [clojure.core$reduce invokeStatic "core.clj" 6544]
  [clojure.core$mapv invokeStatic "core.clj" 6618]
  [clojure.core$mapv invoke "core.clj" 6618]
  [internal.node$pull_input_values invokeStatic "node.clj" 1107]
  [internal.node$pull_input_values invoke "node.clj" 1104]
  [editor.game_project$GameProjectNode$behavior$build_targets invokeStatic "game_project.clj" 207]
  [editor.game_project$GameProjectNode$behavior$build_targets invoke "game_project.clj" 207]
  [clojure.lang.Var invoke "Var.java" 383]
  [internal.node.NodeImpl produce_value "node.clj" 268]
  [internal.node$node_value_STAR_ invokeStatic "node.clj" 335]
  [internal.node$node_value_STAR_ invoke "node.clj" 330]
  [internal.node$node_value invokeStatic "node.clj" 357]
  [internal.node$node_value invoke "node.clj" 350]
  [dynamo.graph$node_value invokeStatic "graph.clj" 742]
  [dynamo.graph$node_value invoke "graph.clj" 712]
  [editor.defold_project$build invokeStatic "defold_project.clj" 306]
  [editor.defold_project$build invoke "defold_project.clj" 300]
  [editor.defold_project$build_and_write invokeStatic "defold_project.clj" 368]
  [editor.defold_project$build_and_write invoke "defold_project.clj" 358]
  [editor.defold_project$build_and_save_project$fn__31456 invoke "defold_project.clj" 587]
  [clojure.core$binding_conveyor_fn$fn__4676 invoke "core.clj" 1938]
  [clojure.lang.AFn call "AFn.java" 18]
  [java.util.concurrent.FutureTask run "FutureTask.java" 266]
  [java.util.concurrent.ThreadPoolExecutor runWorker "ThreadPoolExecutor.java" 1142]
  [java.util.concurrent.ThreadPoolExecutor$Worker run "ThreadPoolExecutor.java" 617]
  [java.lang.Thread run "Thread.java" 745]]}
```

This is a cyclic dependency which means that you for instance have a collection containing a collection proxy or factory which directly or indirectly references itself, thus creating a dependency loop.

--- a/editor/src/clj/editor/app_view.clj
+++ b/editor/src/clj/editor/app_view.clj
@@ -800,12 +800,17 @@
   (let [basis (g/now)
         proj-path (some-> (owning-resource basis node-id) resource/proj-path)
         resource-path (or proj-path "'unknown'")]
-    (case node-type-name
-      "editor.collection/CollectionNode" (format "This is caused by a two or more collections referenced in a loop from either collection proxies or collection factories. One of the involved resources is: %s." resource-path)
-      "editor.game-object/GameObjectNode" (format "This is caused by two or more game objects referenced in a loop from game object factories. One of the involved resources is: %s." resource-path)
-      "editor.code.script/ScriptNode" (format "This is caused by two or more Lua modules required in a loop. One of the involved resources is: %s." resource-path)
-      (format "This is caused by two or more resources referenced in a loop. One of the involved resources is: %s." resource-path))))
+    (case (g/node-type-kw basis node-id)
+      :editor.collection/CollectionNode
+      (format "This is caused by a two or more collections referenced in a loop from either collection proxies or collection factories. One of the involved resources is: %s." resource-path)
 
+      :editor.game-object/GameObjectNode
+      (format "This is caused by two or more game objects referenced in a loop from game object factories. One of the involved resources is: %s." resource-path)
+
+      :editor.code.script/ScriptNode
+      (format "This is caused by two or more Lua modules required in a loop. One of the involved resources is: %s." resource-path)
+
+      (format "This is caused by two or more resources referenced in a loop. One of the involved resources is: %s." resource-path))))
 (defn- build-project!
   [project evaluation-context extra-build-targets old-artifact-map render-progress!]
   (let [game-project (project/get-resource-node project "/game.project" evaluation-context)

--- a/editor/src/clj/editor/app_view.clj
+++ b/editor/src/clj/editor/app_view.clj
@@ -29,6 +29,7 @@
             [editor.changes-view :as changes-view]
             [editor.code.data :refer [CursorRange->line-number]]
             [editor.console :as console]
+            [editor.core :as core]
             [editor.debug-view :as debug-view]
             [editor.defold-project :as project]
             [editor.dialogs :as dialogs]
@@ -790,6 +791,22 @@
                                 :text "If the engine is already running, shut down the process manually and retry"}]}
            :content (.getMessage e)})))))
 
+(defn- owning-resource [basis node-id]
+  (when-some [owning-resource-node-id (core/scope-of-type basis node-id resource-node/ResourceNode)]
+             [owning-resource (resource-node/resource basis owning-resource-node-id)]
+    owning-resource))
+
+(defn- get-cycle-detected-help-message
+  [node-type-name node-id]
+  (let [basis (g/now)
+        proj-path (some-> (owning-resource basis node-id) resource/proj-path)
+        resource-path (some? proj-path) proj-path "'unknown'"]
+    (case node-type-name
+      "editor.collection/CollectionNode" (format "This is caused by a two or more collections referenced in a loop from either collection proxies or collection factories. One of the involved resources is: %s." resource-path)
+      "editor.game-object/GameObjectNode" (format "This is caused by two or more game objects referenced in a loop from game object factories. One of the involved resources is: %s." resource-path)
+      "editor.code.script/ScriptNode" (format "This is caused by two or more Lua modules required in a loop. One of the involved resources is: %s." resource-path)
+      (format "This is caused by two or more resources referenced in a loop. One of the involved resources is: %s." resource-path))))
+
 (defn- build-project!
   [project evaluation-context extra-build-targets old-artifact-map render-progress!]
   (let [game-project (project/get-resource-node project "/game.project" evaluation-context)
@@ -797,6 +814,17 @@
     (try
       (ui/with-progress [render-progress! render-progress!]
         (build/build-project! project game-project evaluation-context extra-build-targets old-artifact-map render-progress!))
+      (catch clojure.lang.ExceptionInfo e
+        (if (= "cycle-detected" (-> e ex-data :cause))
+          (ui/run-later
+            (dialogs/make-info-dialog
+              {:title "Build Error"
+               :icon :icon/triangle-error
+               :header "Cyclic resource dependency detected"
+               :content {:fx/type fxui/label
+                         :style-class "dialog-content-padding"
+                         :text (get-cycle-detected-help-message (-> e ex-data :node-type-name) (-> e ex-data :node-id))}}))
+          (error-reporting/report-exception! e)))
       (catch Throwable error
         (error-reporting/report-exception! error)
         nil))))

--- a/editor/src/clj/editor/app_view.clj
+++ b/editor/src/clj/editor/app_view.clj
@@ -800,7 +800,7 @@
   [node-type-name node-id]
   (let [basis (g/now)
         proj-path (some-> (owning-resource basis node-id) resource/proj-path)
-        resource-path (some? proj-path) proj-path "'unknown'"]
+        resource-path (or proj-path "'unknown'")]
     (case node-type-name
       "editor.collection/CollectionNode" (format "This is caused by a two or more collections referenced in a loop from either collection proxies or collection factories. One of the involved resources is: %s." resource-path)
       "editor.game-object/GameObjectNode" (format "This is caused by two or more game objects referenced in a loop from game object factories. One of the involved resources is: %s." resource-path)

--- a/editor/src/clj/editor/app_view.clj
+++ b/editor/src/clj/editor/app_view.clj
@@ -796,7 +796,7 @@
     (resource-node/resource basis owning-resource-node-id)))
 
 (defn- get-cycle-detected-help-message
-  [node-type-name node-id]
+  [node-id]
   (let [basis (g/now)
         proj-path (some-> (owning-resource basis node-id) resource/proj-path)
         resource-path (or proj-path "'unknown'")]

--- a/editor/src/clj/editor/app_view.clj
+++ b/editor/src/clj/editor/app_view.clj
@@ -811,6 +811,7 @@
       (format "This is caused by two or more Lua modules required in a loop. One of the involved resources is: %s." resource-path)
 
       (format "This is caused by two or more resources referenced in a loop. One of the involved resources is: %s." resource-path))))
+
 (defn- build-project!
   [project evaluation-context extra-build-targets old-artifact-map render-progress!]
   (let [game-project (project/get-resource-node project "/game.project" evaluation-context)

--- a/editor/src/clj/editor/app_view.clj
+++ b/editor/src/clj/editor/app_view.clj
@@ -793,8 +793,7 @@
 
 (defn- owning-resource [basis node-id]
   (when-some [owning-resource-node-id (core/scope-of-type basis node-id resource-node/ResourceNode)]
-             [owning-resource (resource-node/resource basis owning-resource-node-id)]
-    owning-resource))
+    (resource-node/resource basis owning-resource-node-id)))
 
 (defn- get-cycle-detected-help-message
   [node-type-name node-id]

--- a/editor/src/clj/editor/app_view.clj
+++ b/editor/src/clj/editor/app_view.clj
@@ -827,7 +827,7 @@
                :header "Cyclic resource dependency detected"
                :content {:fx/type fxui/label
                          :style-class "dialog-content-padding"
-                         :text (get-cycle-detected-help-message (-> e ex-data :node-type-name) (-> e ex-data :node-id))}}))
+                         :text (get-cycle-detected-help-message (-> e ex-data :node-id))}}))
           (error-reporting/report-exception! e)))
       (catch Throwable error
         (error-reporting/report-exception! error)

--- a/editor/src/clj/editor/app_view.clj
+++ b/editor/src/clj/editor/app_view.clj
@@ -798,15 +798,15 @@
         resource-path (or proj-path "'unknown'")]
     (case (g/node-type-kw basis node-id)
       :editor.collection/CollectionNode
-      (format "This is caused by a two or more collections referenced in a loop from either collection proxies or collection factories. One of the involved resources is: %s." resource-path)
+      (format "This is caused by a two or more collections referenced in a loop from either collection proxies or collection factories. One of the resources is: %s." resource-path)
 
       :editor.game-object/GameObjectNode
-      (format "This is caused by two or more game objects referenced in a loop from game object factories. One of the involved resources is: %s." resource-path)
+      (format "This is caused by two or more game objects referenced in a loop from game object factories. One of the resources is: %s." resource-path)
 
       :editor.code.script/ScriptNode
-      (format "This is caused by two or more Lua modules required in a loop. One of the involved resources is: %s." resource-path)
+      (format "This is caused by two or more Lua modules required in a loop. One of the resources is: %s." resource-path)
 
-      (format "This is caused by two or more resources referenced in a loop. One of the involved resources is: %s." resource-path))))
+      (format "This is caused by two or more resources referenced in a loop. One of the resources is: %s." resource-path))))
 
 (defn- build-project!
   [project evaluation-context extra-build-targets old-artifact-map render-progress!]

--- a/editor/src/clj/editor/app_view.clj
+++ b/editor/src/clj/editor/app_view.clj
@@ -814,7 +814,7 @@
       (ui/with-progress [render-progress! render-progress!]
         (build/build-project! project game-project evaluation-context extra-build-targets old-artifact-map render-progress!))
       (catch clojure.lang.ExceptionInfo e
-        (if (= "cycle-detected" (-> e ex-data :cause))
+        (if (= :cycle-detected (-> e ex-data :cause))
           (ui/run-later
             (dialogs/make-info-dialog
               {:title "Build Error"

--- a/editor/src/clj/editor/app_view.clj
+++ b/editor/src/clj/editor/app_view.clj
@@ -29,7 +29,6 @@
             [editor.changes-view :as changes-view]
             [editor.code.data :refer [CursorRange->line-number]]
             [editor.console :as console]
-            [editor.core :as core]
             [editor.debug-view :as debug-view]
             [editor.defold-project :as project]
             [editor.dialogs :as dialogs]
@@ -75,13 +74,15 @@
             [editor.url :as url]
             [editor.view :as view]
             [editor.workspace :as workspace]
+            [internal.graph.types :as gt]
             [internal.util :refer [first-where]]
             [service.log :as log]
             [util.http-server :as http-server]
             [util.profiler :as profiler]
             [util.thread-util :as thread-util]
             [service.smoke-log :as slog])
-  (:import [com.defold.editor Editor]
+  (:import [clojure.lang ExceptionInfo]
+           [com.defold.editor Editor]
            [com.defold.editor UIUtil]
            [com.sun.javafx.scene NodeHelper]
            [java.io BufferedReader File IOException]
@@ -791,14 +792,9 @@
                                 :text "If the engine is already running, shut down the process manually and retry"}]}
            :content (.getMessage e)})))))
 
-(defn- owning-resource [basis node-id]
-  (when-some [owning-resource-node-id (core/scope-of-type basis node-id resource-node/ResourceNode)]
-    (resource-node/resource basis owning-resource-node-id)))
-
-(defn- get-cycle-detected-help-message
-  [node-id]
+(defn- get-cycle-detected-help-message [node-id]
   (let [basis (g/now)
-        proj-path (some-> (owning-resource basis node-id) resource/proj-path)
+        proj-path (some-> (resource-node/owner-resource basis node-id) resource/proj-path)
         resource-path (or proj-path "'unknown'")]
     (case (g/node-type-kw basis node-id)
       :editor.collection/CollectionNode
@@ -819,7 +815,7 @@
     (try
       (ui/with-progress [render-progress! render-progress!]
         (build/build-project! project game-project evaluation-context extra-build-targets old-artifact-map render-progress!))
-      (catch clojure.lang.ExceptionInfo e
+      (catch ExceptionInfo e
         (if (= :cycle-detected (-> e ex-data :cause))
           (ui/run-later
             (dialogs/make-info-dialog
@@ -828,7 +824,7 @@
                :header "Cyclic resource dependency detected"
                :content {:fx/type fxui/label
                          :style-class "dialog-content-padding"
-                         :text (get-cycle-detected-help-message (-> e ex-data :node-id))}}))
+                         :text (get-cycle-detected-help-message (-> e ex-data :endpoint gt/endpoint-node-id))}}))
           (error-reporting/report-exception! e)))
       (catch Throwable error
         (error-reporting/report-exception! error)

--- a/editor/src/clj/editor/resource_node.clj
+++ b/editor/src/clj/editor/resource_node.clj
@@ -162,6 +162,52 @@
     (and (g/error? value)
          (g/error-fatal? value))))
 
+(defn owner-resource-node
+  ([node-id]
+   (owner-resource-node (g/now) node-id))
+  ([basis node-id]
+   ;; The owner resource node is the first non-override ResourceNode we find by
+   ;; traversing the explicit :cascade-delete connections between nodes. I.e, we
+   ;; want to find the ResourceNode that will delete our node if the resource is
+   ;; deleted from the project.
+   (let [node (g/node-by-id basis node-id)]
+     ;; Embedded resources will be represented by ResourceNodes, but we want the
+     ;; ultimate owner of the embedded resource, so we keep looking if we
+     ;; encounter a d that is an override node.
+     (if (and (nil? (gt/original node))
+              (g/node-instance*? ResourceNode node))
+
+       ;; We found our owner ResourceNode. Return it.
+       node
+
+       ;; This is not our owner ResourceNode. Recursively follow the outgoing
+       ;; connections that connect to a :cascade-delete input.
+       (let [outgoing-arcs (gt/arcs-by-source basis node-id)
+             outgoing-arcs-by-target-id (group-by gt/target-id outgoing-arcs)]
+         (some (fn [[target-id outgoing-arcs]]
+                 (let [target-node-type (g/node-type* basis target-id)
+                       cascade-delete-input-labels (g/cascade-deletes target-node-type)
+                       connects-to-cascade-delete-input? (comp cascade-delete-input-labels gt/target-label)]
+                   (some (fn [outgoing-arc]
+                           (when (connects-to-cascade-delete-input? outgoing-arc)
+                             (owner-resource-node basis target-id)))
+                         outgoing-arcs)))
+               outgoing-arcs-by-target-id))))))
+
+(defn owner-resource-node-id
+  ([node-id]
+   (owner-resource-node-id (g/now) node-id))
+  ([basis node-id]
+   (some->> (owner-resource-node basis node-id)
+            (gt/node-id))))
+
+(defn owner-resource
+  ([node-id]
+   (owner-resource (g/now) node-id))
+  ([basis node-id]
+   (some->> (owner-resource-node basis node-id)
+            (resource-node-resource basis))))
+
 (defn make-ddf-dependencies-fn [ddf-type]
   (fn [source-value]
     (into []

--- a/editor/src/clj/internal/node.clj
+++ b/editor/src/clj/internal/node.clj
@@ -1335,11 +1335,8 @@
 
 (defn mark-in-production [node-id label evaluation-context]
   (if (contains? (:in-production evaluation-context) (gt/endpoint node-id label))
-    (throw (ex-info (format "Cycle detected on node type %s and output %s"
-                      (node-type-name node-id evaluation-context)
-                      label)
+    (throw (ex-info "Cycle detected on node"
                     {:cause :cycle-detected
-                     :node-type-name (node-type-name node-id evaluation-context)
                      :node-id node-id}))
     (update evaluation-context :in-production conj (gt/endpoint node-id label))))
 

--- a/editor/src/clj/internal/node.clj
+++ b/editor/src/clj/internal/node.clj
@@ -1338,7 +1338,7 @@
     (throw (ex-info (format "Cycle detected on node type %s and output %s"
                       (node-type-name node-id evaluation-context)
                       label)
-                    {:cause "cycle-detected"
+                    {:cause :cycle-detected
                      :node-type-name (node-type-name node-id evaluation-context)
                      :node-id node-id}))
     (update evaluation-context :in-production conj (gt/endpoint node-id label))))

--- a/editor/src/clj/internal/node.clj
+++ b/editor/src/clj/internal/node.clj
@@ -1333,12 +1333,16 @@
         node (gt/node-by-id-at basis node-id)]
     (type-name (gt/node-type node))))
 
-(defn mark-in-production [node-id label evaluation-context]
-  (if (contains? (:in-production evaluation-context) (gt/endpoint node-id label))
+(defn- update-in-production [in-production endpoint]
+  (if (contains? in-production endpoint)
     (throw (ex-info "Cycle detected on node"
                     {:cause :cycle-detected
-                     :node-id node-id}))
-    (update evaluation-context :in-production conj (gt/endpoint node-id label))))
+                     :endpoint endpoint
+                     :in-production in-production}))
+    (conj in-production endpoint)))
+
+(defn mark-in-production [node-id label evaluation-context]
+  (update evaluation-context :in-production update-in-production (gt/endpoint node-id label)))
 
 (defn- mark-in-production-form [node-id-sym label-sym evaluation-context-sym forms]
   `(let [~evaluation-context-sym (mark-in-production ~node-id-sym ~label-sym ~evaluation-context-sym)]

--- a/editor/src/clj/internal/node.clj
+++ b/editor/src/clj/internal/node.clj
@@ -1334,11 +1334,14 @@
     (type-name (gt/node-type node))))
 
 (defn mark-in-production [node-id label evaluation-context]
-  (assert (not (contains? (:in-production evaluation-context) (gt/endpoint node-id label)))
-          (format "Cycle detected on node type %s and output %s"
-                  (node-type-name node-id evaluation-context)
-                  label))
-  (update evaluation-context :in-production conj (gt/endpoint node-id label)))
+  (if (contains? (:in-production evaluation-context) (gt/endpoint node-id label))
+    (throw (ex-info (format "Cycle detected on node type %s and output %s"
+                      (node-type-name node-id evaluation-context)
+                      label)
+                    {:cause "cycle-detected"
+                     :node-type-name (node-type-name node-id evaluation-context)
+                     :node-id node-id}))
+    (update evaluation-context :in-production conj (gt/endpoint node-id label))))
 
 (defn- mark-in-production-form [node-id-sym label-sym evaluation-context-sym forms]
   `(let [~evaluation-context-sym (mark-in-production ~node-id-sym ~label-sym ~evaluation-context-sym)]

--- a/editor/test/internal/node_test.clj
+++ b/editor/test/internal/node_test.clj
@@ -324,24 +324,24 @@
   (testing "output dependent on itself"
     (with-clean-system
       (let [[node] (tx-nodes (g/make-node world DependencyNode))]
-        (is (thrown? AssertionError (g/node-value node :out-from-self))))))
+        (is (thrown? ExceptionInfo (g/node-value node :out-from-self))))))
   (testing "output dependent on itself connected to downstream input"
     (with-clean-system
       (let [[node0 node1] (tx-nodes (g/make-node world DependencyNode) (g/make-node world DependencyNode))]
         (g/transact
          (g/connect node0 :out-from-self node1 :in))
-        (is (thrown? AssertionError (g/node-value node1 :out-from-in))))))
+        (is (thrown? ExceptionInfo (g/node-value node1 :out-from-in))))))
   (testing "cycle of period 1"
     (with-clean-system
       (let [[node] (tx-nodes (g/make-node world DependencyNode))]
         (g/transact (g/connect node :out-from-in node :in))
-        (is (thrown? AssertionError (g/node-value node :out-from-in))))))
+        (is (thrown? ExceptionInfo (g/node-value node :out-from-in))))))
   (testing "cycle of period 2 (single transaction)"
     (with-clean-system
       (let [[node0 node1] (tx-nodes (g/make-node world DependencyNode) (g/make-node world DependencyNode))]
         (g/transact [(g/connect node0 :out-from-in node1 :in)
                      (g/connect node1 :out-from-in node0 :in)])
-        (is (thrown? AssertionError (g/node-value node1 :out-from-in)))))))
+        (is (thrown? ExceptionInfo (g/node-value node1 :out-from-in)))))))
 
 (g/defnode BasicNode
   (input basic-input g/Int)

